FeedStoreChallenge/FeedStore/CoreData/CoreDataFeedStore.swift

```
@@ -0,0 +1,125 @@
+        do {
+            if let currentCache = try CDCache.fetchCachedFeed(managedContext) {
+                managedContext.delete(currentCache)
+            }
+        } catch { /* If there is an error then just insert the feed in Core Data*/ }
+        
+        let cdFeed = mapLocalFeedToCoreDataFeed(feed, timestamp: timestamp)
+        
+        if let saveError = saveContext() {
+            completion(.some(saveError))
+        } else {
+            completion(.none)
+        }
+    }
+    
+    public func retrieve(completion: @escaping RetrievalCompletion) {
+        do {
+            if let cache = try CDCache.fetchCachedFeed(managedContext) {
+                let imageFeed = cache.feed.compactMap { ($0 as? CDFeedImage)?.local }
+                completion(.found(feed: imageFeed, timestamp: cache.timestamp))
+            } else {
+                completion(.empty)
+            }
+        } catch {
+            completion(.failure(error))
+        }
+    }
+    
+    
+    
+    // MARK: Helper Methods
+    
+    private func saveContext() -> Error? {
+        if storeContainer.viewContext.hasChanges {
+            do {
+                try storeContainer.viewContext.save()
+                return nil
+            } catch {
+                return error
+            }
+        } else {
+            return nil
+        }
+    }
+    
+    private func mapLocalFeedToCoreDataFeed(_ feed: [LocalFeedImage], timestamp: Date) -> CDCache {
+        let cache = CDCache(context: managedContext)
+        cache.feed = CDFeedImage.mapLocalFeedImagesToCoreDataFeedImages(from: feed, in: managedContext)
+        cache.timestamp = timestamp
+        return cache
+    }
+    
+    private func deleteCache() throws {
+        do {
+            if let currentCache = try CDCache.fetchCachedFeed(managedContext) {
+                managedContext.delete(currentCache)
+            }
+        } catch { /* If there is an error then just insert the feed in Core Data*/ }
```

Errors should be handled. Remove the do/catch and let it propagate.